python/cudf_polars/cudf_polars/utils/dtypes.py

```diff
@@ -61,12 +71,57 @@ def can_cast(from_: plc.DataType, to: plc.DataType) -> bool:
     True if casting is supported, False otherwise
     """
     return (
-        plc.traits.is_fixed_width(to)
-        and plc.traits.is_fixed_width(from_)
-        and plc.unary.is_supported_cast(from_, to)
+        (
+            plc.traits.is_fixed_width(to)
+            and plc.traits.is_fixed_width(from_)
+            and plc.unary.is_supported_cast(from_, to)
+        )
+        or (from_.id() == plc.TypeId.STRING and is_numeric_not_bool(to))
+        or (to.id() == plc.TypeId.STRING and is_numeric_not_bool(from_))
     )
 
 
+def is_order_preserving_cast(from_: plc.DataType, to: plc.DataType) -> bool:
+    """
+    Determine if a cast would preserve the order of the source data.
+
+    Parameters
+    ----------
+    from_
+        Source datatype
+    to
+        Target datatype
+
+    Returns
+    -------
+    True if the cast is order-preserving, False otherwise
+    """
+    if from_.id() == to.id():
+        return True
+
+    if is_integral_not_bool(from_) and is_integral_not_bool(to):
+        # True if signedness is the same and the target is larger
+        if plc.traits.is_unsigned(from_) == plc.traits.is_unsigned(to):
+            if plc.types.size_of(to) >= plc.types.size_of(from_):
+                return True
+        elif (plc.traits.is_unsigned(from_) and not plc.traits.is_unsigned(to)) and (
+            plc.types.size_of(to) > plc.types.size_of(from_)
+        ):
+            # Unsigned to signed is order preserving if target is large enough
+            # But signed to unsigned is never order preserving due to negative values
+            return True
+    elif is_floating_point(from_) and is_floating_point(to):
+        # True if the target is larger
+        if plc.types.size_of(to) > plc.types.size_of(from_):
+            return True
+    elif (is_integral_not_bool(from_) and is_floating_point(to)) and (
+        plc.types.size_of(to) > plc.types.size_of(from_)
+    ):
+        # Int64 fits in float64, but not in float32
+        return True
```

Comment:
I'm not sure about this one, the range of int64 is contained in the range of float32. I believe the mapping preserves the order.

Also, what about float to integral? I suppose it depends on what happens to the out of bounds values.

Reply:
Correct the float to integral cases all fall into the last `return False` because the out of range values might lose their ordering in the cast.

Reply:
Why? If the range is just clamped, then you have no problem, ordering is preserved.

Reply:
Because I expect the equivalent of this:
```
>>> ary = np.array([1, 2, float(np.iinfo('int64').max) + 1])
>>> ary
array([1.00000000e+00, 2.00000000e+00, 9.22337204e+18] # ordered
>>> ary.astype('int64')
array([1, 2, -9223372036854775808]) # not ordered
```

Reply:
Is that what cudf does though?

Reply:
With `wrap_numerical=True` in the cast for polars, it clamps, AFAICT

Reply:
no, libcudf clamps. If we're ok encoding libcudf specific implementation behavior into this function we could pass the float to int cases.

Reply:
@wence- is this current behavior considered buggy?

```python
>>> df.collect()
shape: (3, 1)
┌───────────┐
│ a         │
│ ---       │
│ f32       │
╞═══════════╡
│ 1.0       │
│ 2.0       │
│ 9.2234e18 │
└───────────┘
>>> df.select(pl.col('a').cast(pl.Int64)).collect()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/envs/cudf_dev/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 2055, in collect
    return wrap_df(ldf.collect(callback))
                   ^^^^^^^^^^^^^^^^^^^^^
polars.exceptions.InvalidOperationError: conversion from `f32` to `i64` failed in column 'a' for 1 out of 3 values: [9.2234e18]
>>> df.select(pl.col('a').cast(pl.Int64)).collect(engine=pl.GPUEngine())
shape: (3, 1)
┌─────────────────────┐
│ a                   │
│ ---                 │
│ i64                 │
╞═════════════════════╡
│ 1                   │
│ 2                   │
│ 9223372036854775807 │
└─────────────────────┘
```

Reply:
Probably yes, we should barf for the cases where we're strictish mode because we haven't implemented those

Reply:
The issue I forsee here is that `wrap_numerical=False`, `strict=True` is the default. This means that by default the GPU backend will also have to scan during the float-int cast for the presence of these values and throw. This is shaping up to be a pattern that occurs in several places within the codebase, and it's probably not ideal to need to scan before every cast. 

For now I have passed the float-int conversions through this function which retains the existing behavior, since regardless of if OOB values are nullified or clamped we'll retain order. I will raise a separate issue to discuss the proliferation of scanning as a result of polars defaults.